app/helpers/automated_tests_client_helper.rb

```diff
@@ -415,38 +407,33 @@ def self.perform(host_with_port, test_scripts, user_api_key, server_api_key, gro
         end
       rescue Exception => e
         submission = submission_id.nil? ? nil : Submission.find(submission_id)
-        create_all_test_scripts_error_result(test_scripts, assignment, grouping, submission,
+        requested_by = User.find_by(api_key: user_api_key)
+        create_all_test_scripts_error_result(test_scripts, assignment, grouping, submission, requested_by,
                                              I18n.t('automated_tests.test_result.all_tests'),
                                              I18n.t('automated_tests.test_result.no_server_connection',
                                                     {hostname: test_server_host, error: e.message}))
       end
     end
   end
 
-  def self.process_test_result(raw_result, test_scripts_ran, assignment, grouping, submission)
+  def self.process_test_result(raw_result, test_scripts_ran, assignment, grouping, submission, requested_by)
 
     # check that results are somewhat well-formed xml at the top level (i.e. they don't crash the parser)
     result = nil
     begin
       result = Hash.from_xml(raw_result)
     rescue => e
-      create_all_test_scripts_error_result(test_scripts_ran, assignment, grouping, submission,
+      create_all_test_scripts_error_result(test_scripts_ran, assignment, grouping, submission, requested_by,
                                            I18n.t('automated_tests.test_result.all_tests'),
                                            I18n.t('automated_tests.test_result.bad_results', {xml: e.message}))
-      unless submission.nil?
-        submission.set_marks_for_tests

```

Comment:
In case of errors it's directly in the function create_all_test_scripts_error_result, so it was a duplicate. It's still at the end of process_test_result if no errors happen.
(the diff makes it hard to see it)